src/rmt.rs

```diff
@@ -856,3 +915,189 @@ mod chip {
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use core::ptr;
+
+pub struct RxRmtDriver<'d> {
+    channel: u8,
+    _p: PhantomData<&'d mut ()>,
+    pub pulse_pair_vec: Vec<PulsePair>,
+}
+
+impl<'d> RxRmtDriver<'d> {
+    /// Initialise the rmt module with the specified pin, channel and configuration.
+    ///
+    /// To uninstall the driver just drop it.
+    ///
+    /// Internally this calls `rmt_config()` and `rmt_driver_install()`.
+
+    pub fn new<C: RmtChannel>(
+        _channel: impl Peripheral<P = C> + 'd,
+        pin: impl Peripheral<P = impl InputPin> + 'd,
+        config: &ReceiveConfig,
+        rx_buffer_size_in_bytes: u32,
+    ) -> Result<Self, EspError> {
+        crate::into_ref!(pin);
+
+        let flags = 0;
+
+        #[cfg(not(any(esp32, esp32c2)))]
+        let carrier_en = config.carrier.is_some();
+
+        #[cfg(not(any(esp32, esp32c2)))]
+        let carrier = config.carrier.unwrap_or_default();
+
+        let sys_config = rmt_config_t {
+            rmt_mode: rmt_mode_t_RMT_MODE_RX,
+            channel: C::channel(),
+            gpio_num: pin.pin(),
+            clk_div: config.clock_divider,
+            mem_block_num: config.mem_block_num,
+            flags,
+            __bindgen_anon_1: rmt_config_t__bindgen_ty_1 {
+                rx_config: rmt_rx_config_t {
+                    idle_threshold: config.idle_threshold,
+                    filter_ticks_thresh: config.filter_ticks_thresh,
+                    filter_en: config.filter_en,
+                    #[cfg(not(any(esp32, esp32c2)))]
+                    rm_carrier: carrier_en,
+                    #[cfg(not(any(esp32, esp32c2)))]
+                    carrier_freq_hz: carrier.frequency.into(),
+                    #[cfg(not(any(esp32, esp32c2)))]
+                    carrier_level: carrier.carrier_level as u32,
+                    #[cfg(not(any(esp32, esp32c2)))]
+                    carrier_duty_percent: carrier.duty_percent.0,
+                },
+            },
+        };
+
+        let pulse_pair_size: u32 = rx_buffer_size_in_bytes / 4;
+        let pulse_pair_vec: Vec<PulsePair> =
+            Vec::with_capacity(pulse_pair_size.try_into().unwrap());
+
+        unsafe {
+            esp!(rmt_config(&sys_config))?;
+            esp!(rmt_driver_install(C::channel(), rx_buffer_size_in_bytes, 0))?;
+        }
+
+        Ok(Self {
+            channel: C::channel() as _,
+            _p: PhantomData,
+            pulse_pair_vec,
+        })
+    }
+
+    pub fn channel(&self) -> rmt_channel_t {
+        self.channel as _
+    }
+
+    /// Start receiving
+    pub fn start(&self) -> Result<(), EspError> {
+        esp!(unsafe { rmt_rx_start(self.channel(), true) })
+    }
+
+    /// Stop receiving
+    pub fn stop(&self) -> Result<(), EspError> {
+        esp!(unsafe { rmt_rx_stop(self.channel()) })
+    }
+
+    // Set ticks_to_wait to 0 for non-blocking.
+    pub fn get_rmt_items(&mut self, ticks_to_wait: u32) -> Result<u32, EspError> {
```

Comment:
1. Rename to simply `receive`
2. `ticks_to_wait` should be of `TickType_t` type. See also the `delay` module. This way the user can e.g. pass `delay::BLOCK` here.
3. Method should take a `& mut [PulsePair]` buffer reference
4. Method should return `usize`, not `u32`, which would be the number of items fetched from the ring buffer and written back to the user-supplied buffer